src/components/Identity/createComponent.js

```diff
@@ -0,0 +1,179 @@
+import createCustomerIds from "./customerIds/createCustomerIds";
+import { defer } from "../../utils";
+import { EXPERIENCE_CLOUD_ID } from "./constants/cookieNames";
+import migration from "./migration";
+import createEvent from "../DataCollector/createEvent";
+
+const addIdsContext = (payload, ecid) => {
+  payload.addIdentity(EXPERIENCE_CLOUD_ID, {
+    id: ecid
+  });
+};
+
+export default (idSyncs, manualIdSyncs, cookieJar, config, logger, network) => {
+  let optIn;
+  let deferredForEcid;
+  let lifecycle;
+  let customerIds;
+  let alreadyQueriedForIdSyncs = false;
+  const { migrateIds, imsOrgId } = config;
+
+  const createPayload = event => {
+    const payload = network.createPayload();
+    payload.addEvent(event);
+    return payload;
+  };
+
+  const readEcid = () => {
+    let ecid = cookieJar.get(EXPERIENCE_CLOUD_ID);
+    if (!ecid && migrateIds) {
+      ecid = migration(imsOrgId).readEcidFromAmcvCookie();
+      if (ecid) {
+        cookieJar.set(EXPERIENCE_CLOUD_ID, ecid);
+      }
+    }
+    return ecid;
+  };
+
+  const makeServerCall = payload => {
+    return lifecycle.onBeforeDataCollection({ payload }).then(() => {
+      return network.sendRequest(payload, payload.expectsResponse);
+    });
+  };
+
+  const createEcidEvent = () => {
+    const event = createEvent();
+    const payload = createPayload(event);
+
+    return lifecycle
+      .onBeforeEvent({
+        event,
+        options: {},
+        isViewStart: false,
+        documentUnloading: false
+      }) // FIXME: We shouldn't need an event.
+      .then(() => optIn.whenOptedIn())
+      .then(() => makeServerCall(payload));
+  };
+
+  const getEcid = () => {
+    const ecid = readEcid();
+    if (ecid) {
+      return ecid;
+    }
+    return createEcidEvent();
+  };
```

Comment:
It looks like this function returns an ECID value sometimes and at other times returns a promise. This code: 

```
config.reactorRegisterGetEcid(() => {
  return optIn.isOptedIn() ? getEcid() : undefined;
});
```

provides a function to the Launch extension that allows the extension to access the ECID immediately. It isn't set up to handle the case where a promise is returned. I think in this case we need to change the code to use `readEcid` instead:

```
config.reactorRegisterGetEcid(() => {
  return optIn.isOptedIn() ? readEcid() : undefined;
});
```

Reply:
I see this function is what's being exposed as the `getEcid` command. What if a user does not yet have an ECID and the customer does this:

```
alloy("getEcid");
alloy("getEcid");
```

Could this end up generating two ECIDs for the user? What about if a user did this:

```
alloy("event");
alloy("getEcid");
```

Reply:
I agree with @Aaronius. I think having a function that returns 2 different types is weird in general, but we might need to add guards against multiple calls in the same session.